packages/pg-protocol/src/serializer.ts

```diff
@@ -106,50 +106,53 @@ type BindOpts = {
   binary?: boolean
   statement?: string
   values?: any[]
+  // optional map from JS value to postgres value per parameter
+  valueMap?: (param: any) => any
 }
 
-const bind = (config: BindOpts = {}): Buffer => {
-  // normalize config
-  const portal = config.portal || ''
-  const statement = config.statement || ''
-  const binary = config.binary || false
-  var values = config.values || emptyArray
-  var len = values.length
+const paramWriter = new Writer()
 
-  var useBinary = false
-  // TODO(bmc): all the loops in here aren't nice, we can do better
-  for (var j = 0; j < len; j++) {
-    useBinary = useBinary || values[j] instanceof Buffer
-  }
-
-  var buffer = writer.addCString(portal).addCString(statement)
-  if (!useBinary) {
-    buffer.addInt16(0)
-  } else {
-    buffer.addInt16(len)
-    for (j = 0; j < len; j++) {
-      buffer.addInt16(values[j] instanceof Buffer ? 1 : 0)
-    }
-  }
-  buffer.addInt16(len)
-  for (var i = 0; i < len; i++) {
+const writeValues = function (values: any[], valueMap?: (val: any) => any): void {
+  for (let i = 0; i < values.length; i++) {
     var val = values[i]
     if (val === null || typeof val === 'undefined') {
-      buffer.addInt32(-1)
+      writer.addInt16(0)
+      paramWriter.addInt32(-1)
     } else if (val instanceof Buffer) {
-      buffer.addInt32(val.length)
-      buffer.add(val)
+      writer.addInt16(1)
+      const mappedVal = valueMap ? valueMap(val) : val
+      paramWriter.addInt32(mappedVal.length)
+      paramWriter.add(mappedVal)
     } else {
-      buffer.addInt32(Buffer.byteLength(val))
-      buffer.addString(val)
+      writer.addInt16(0)
+      const mappedVal = valueMap ? valueMap(val) : val
+      paramWriter.addInt32(Buffer.byteLength(mappedVal))
+      paramWriter.addString(mappedVal)
```

Comment:
My read of this is that if a Buffer is supplied as an arg, it will be transformed and the result will be expected to still be a Buffer. Ditto for non-Buffers (expected to be string-like).

Instead, how about adding `const mappedVal = valueMap ? valueMap(val, i) : val` prior to the if-else-if-else block? That would allow transforming null or undefined into something else on a per-field basis. It also would allow the transform to return null, a string, or Buffer on a per-field basis and transform from a Buffer type bound as binary to a string one bound as text.

Reply:
Great ideas. Will do, and write a couple more tests covering it too.